nautobot/extras/forms.py

```diff
@@ -227,33 +228,78 @@ def _append_relationships(self):
+
+        - For TYPE_ONE_TO_MANY and TYPE_ONE_TO_ONE relations, if the form's object is on the "source" side of
+          the relationship, verify that the requested "destination" object(s) do not already have any existing
+          RelationshipAssociation to a different source object.
+        - For TYPE_ONE_TO_ONE relations, if the form's object is on the "destination" side of the relationship,
+          verify that the requested "source" object does not have an existing RelationshipAssociation to
+          a different destination object.
+        """
+        for side, relationships in self.instance.get_relationships().items():
+            for relationship in relationships:
+                # If this side of the relationship is a "MANY" relation,
+                # we can coexist with other existing RelationshipAssociations and there's nothing to check now.
+                if relationship.has_many(side):
+                    continue
+
+                peer_side = RelationshipSideChoices.OPPOSITE[side]
+                field_name = f"cr_{relationship.slug}__{peer_side}"
+
+                # Is the form trying to set this field (create/update a RelationshipAssociation(s))?
+                # If not (that is, clearing the field / deleting RelationshipAssociation(s)), we don't need to check.
+                if field_name not in self.cleaned_data or not self.cleaned_data[field_name]:
+                    continue
+
+                # Are any of the objects we want a relationship with already entangled with another object?
+                if relationship.has_many(peer_side):
+                    target_peers = [item for item in self.cleaned_data[field_name]]
+                else:
+                    target_peers = [self.cleaned_data[field_name]]
+                for target_peer in target_peers:
+                    existing_peer_associations = RelationshipAssociation.objects.filter(
+                        relationship=relationship,
+                        **{
+                            f"{peer_side}_id": target_peer.pk,
+                        },
+                    ).exclude(**{f"{side}_id": self.instance.pk})
+                    if existing_peer_associations.count() > 0:
+                        raise ValidationError(
+                            {field_name: f"{target_peer} is already involved in a {relationship} relationship"}
+                        )
+
+        return super().clean()
+
     def _save_relationships(self):
         """Update RelationshipAssociations for all Relationships on form save."""
 
         for field_name in self.relationships:
 
             # Extract the side from the field_name
             # Based on the side, find the list of existing RelationshipAssociation
-            side = field_name.split("__")[-1]
-            peer_side = RelationshipSideChoices.OPPOSITE[side]
+            peer_side = field_name.split("__")[-1]
+            side = RelationshipSideChoices.OPPOSITE[peer_side]
```

Comment:
Curious. Why did these get inverted?

Reply:
This was a change that I made for correctness.

Let's say we have a relationship `"routerid"` with `Device` as `source` and `IPAddress` as `destination`.

With the existing code, when you edit a Device, the form field that gets created on the Device form for this relationship would have been a ModelChoiceField for IPAddress named `cr_routerid__source` - but for this relationship, the *source* is `Device`, the `IPAddress` you select is the *destination*.

So I felt that the form field naming was incorrect and potentially misleading, and swapped the logic around so that the field will instead (IMHO more correctly) be named `cr_routerid__destination`. To keep things working properly, I have to similarly invert the logic here in forms.py that takes the field name and extracts the sides of the relationship out of it.

Hope that makes sense? I'll add a comment or two to the code.